src/iosource/Manager.cc

```diff
@@ -59,124 +109,411 @@ IOSource* Manager::FindSoonest(double* ts)
+	if ( pkt_src && pkt_src->IsOpen() && ! pkt_src->IsLive() && pkt_src->GetNextTimeout() == 0 )
+		ready.insert(pkt_src);
+
+	// Call the appropriate poll method for what's available on the operating system.
+	timer_expired = Poll(ready, timeout, timeout_src);
 
-	if ( soonest_src && (call_count % SELECT_FREQUENCY) != 0 )
-		goto finished;
+	return ready;
+	}
+
+#if defined(HAVE_EPOLL_H)
+
+void Manager::InitQueue()
+	{
+	event_queue = epoll_create1(EPOLL_CLOEXEC);
+	if ( event_queue == -1 )
+		reporter->FatalError("Failed to open epoll() file descriptor: %s", strerror(errno));
 
-	// Select on the join of all file descriptors.
-	fd_set fd_read, fd_write, fd_except;
+	timerfd = timerfd_create(CLOCK_MONOTONIC, 0);
+	if ( timerfd == -1 )
+		reporter->FatalError("Failed to initialize timerfd: %s", strerror(errno));
 
-	FD_ZERO(&fd_read);
-	FD_ZERO(&fd_write);
-	FD_ZERO(&fd_except);
+	epoll_event event;
+	memset(&event, 0, sizeof(epoll_event));
+	event.events = EPOLLIN;
 
-	for ( SourceList::iterator i = sources.begin();
-	      i != sources.end(); ++i )
+	if ( epoll_ctl(event_queue, EPOLL_CTL_ADD, timerfd, &event) != -1 )
 		{
-		Source* src = (*i);
+		DBG_LOG(DBG_MAINLOOP, "Added fd %d from Timerfd", timerfd);
+		events.push_back({});
+		}
+	}
 
-		if ( ! src->src->IsIdle() )
-			// No need to select on sources which we know to
-			// be ready.
-			continue;
+void Manager::RegisterFd(int fd, IOSource* src)
+	{
+	epoll_event event;
+	memset(&event, 0, sizeof(epoll_event));
+	event.events = EPOLLIN;
+	event.data.fd = fd;
 
-		src->Clear();
-		src->src->GetFds(&src->fd_read, &src->fd_write, &src->fd_except);
-		src->SetFds(&fd_read, &fd_write, &fd_except, &maxx);
+	int ret = epoll_ctl(event_queue, EPOLL_CTL_ADD, fd, &event);
+	if ( ret != -1 )
+		{
+		DBG_LOG(DBG_MAINLOOP, "Registered fd %d from %s", fd, src->Tag());
+		events.push_back({});
+		fd_map[fd] = src;
+		Wakeup("RegisterFd");
+		}
+	else
+		{
+		DBG_LOG(DBG_MAINLOOP, "Failed to register fd %d from %s: %s", fd, src->Tag(), strerror(errno));
 		}
+	}
 
-	// We can't block indefinitely even when all sources are dry:
-	// we're doing some IOSource-independent stuff in the main loop,
-	// so we need to return from time to time. (Instead of no time-out
-	// at all, we use a very small one. This lets FreeBSD trigger a
-	// BPF buffer switch on the next read when the hold buffer is empty
-	// while the store buffer isn't filled yet.
+void Manager::UnregisterFd(int fd)
+	{
+	if ( fd_map.find(fd) != fd_map.end() )
+		{
+		int ret = epoll_ctl(event_queue, EPOLL_CTL_DEL, fd, NULL);
+		if ( ret != -1 )
+			DBG_LOG(DBG_MAINLOOP, "Unregistered fd %d", fd);
 
-	struct timeval timeout;
+		events.pop_back();
+		fd_map.erase(fd);
 
-	if ( all_idle )
+		Wakeup("UnregisterFd");
+		}
+	}
+
+bool Manager::Poll(std::set<IOSource*>& ready, double timeout, IOSource* timeout_src)
+	{
+	bool timer_expired = false;
+
+	// TODO: double comparisons are fraught with danger
```

Comment:
Not sure what this TODO is meant to suggest (I know what the challenge with double comparasions is, but not what solution you anticipate for the TODO).